### DISCUSSION TITLE
Upcoming NVIDIA Shield Support

### DISCUSSION THREAD

**Discussion by PRProd** [PRProd/HA-Firemote#47] [General]
# Official NVIDIA Shield Support is coming!
![ShieldRemotesPreview](https://user-images.githubusercontent.com/25486599/225754997-a3672fa1-f094-4115-a714-e4d95a3e8196.jpg)

In the process of this upgrade, much of the code base had to be restructured.  For this reason, I will be asking for help from the community (that's you!) for help with testing before the official release.  The good news is that any FUTURE Android TV additions made to the Firemote project [(vote here)](https://github.com/PRProd/HA-Firemote/discussions/41) should be easier to create and maintain!

The first pre-release will have official support for the NVIDIA Shield TV Pro (2019) since that is the device I own and can test, but if you own a different model, I need your input too!

Here is a list of the NVIDIA Shield models currently available to the public: (let me know if I'm missing one)
 - Shield TV Pro (2019)
 - Shield TV (2019)
 - Shield TV Pro (2017)
 - Shield TV (2017)
 - Shield TV Pro (2015)
 - Shield TV (2015)

## New Features
 - Remote Style Override: Regardless of the kind of device you own, you can switch your Firemote card to look like a completely different remote control.  If you were very clever, you probably knew of a way to do this already, but now it's built in!
 - Visible Device Name Color Choice: As requested in #44, you can now choose a different color for the visible device name text.  As of right now, it should work properly on any Amazon style remote, but it is not yet ready for the Shield version.
 - Sometimes apps made for the Amazon Fire world weren't the same as the apps made for other Android TV style devices which needed to be sorted out.  If you've used Firemote for any non Amazon Fire devices, you may have come across this problem before, as @moh-Brent discovered in #30.  These issues have been all sorted out under the hood which will absolutely result in a better experience

## Currently known limitations:
 - Navigation via the dpad is sllloooowwww.  While I recognize that this is the common experience for anyone who has been using Firemote with a Shield device in 'Strong' compatibility mode, I really want to fix it.  So far, lots of research, no positive results.
 - Support for the visible device name feature on an Shield remote has not been completed, so attempting to use it now will completely kill the layout of an NVIDIA style remote

---
### Beta test with me!
New alpha & beta versions will be occasionally made available on GitHub until I feel strongly enough that the newest version is strong and ready to go.  Keep your eyes on this discussion for updates.  I will post here or reference the discussion thread whenever I push out a new release.

@jrspowers and @MRobi1 have already expressed interest in this upgrade, so I'm tagging them in this discussion post now.

I'm excited to share something with you all very soon!  

- @PRProd  (Doug)

**Comment by PRProd** [comment]
The first beta version is ready for testing!  https://github.com/PRProd/HA-Firemote/releases/tag/v2.0.0-beta1

Please provide feedback!  I appreciate every little bit I can get!

**Reply by jrspowers** [reply]
@PRProd nice man this remote is very nice & works great & all 1 problem that is already noticeable that youtube app launch doesn't work?  & can you also add this app?  vtm go BE & streamz BE & vrt max BE & Tivmate ?  is it also possible to use 6 app launch button would be handy?  thanks in advance for your hard work!

**Reply by jrspowers** [reply]
if I want to turn off the nvidia shield with the remote, does this not work either?

**Comment by MRobi1** [comment]
This is awesome! It's 11pm here right now so I'll play with this tomorrow. 
I've got a few 2019 pro's and some 2017 models as well (I have 9 shields total LOL). There should be no different in the code for the remotes since you can use them interchangeably on either model. I actually have my 2019 pro remote paired to one of my 2017 model shields.

Since I literally just found this project last night through the giant Mushroom thread on the HA boards, I'm unaware of some of the capabilities. Are those quick-launch app buttons user configurable? I assume no since they're likely built into the image but I figure I'd ask since I don't use a single one of those apps even though they're quite common.

**Comment by MRobi1** [comment]
When adding a device name to the bottom of the card, the text starts on the right side and also causes the buttons to move off the remote. Text on top seems to work fine, and of course no text at all is fine.

![image](https://user-images.githubusercontent.com/34860474/225901478-2e27bd70-593e-428f-87cf-2537e090d45a.png)

**Reply by jrspowers** [reply]
he also mentioned that text is not working properly yet!  but does disabling nvidia with the remote work well for you?

**Reply by MRobi1** [reply]
I'm unsure what you mean by "disabling nvidia with the remote"?

**Reply by jrspowers** [reply]
power button not work when off

power button does work at on

I have the compatibility mode : strong (slower) set & everything works there except this above?

**Reply by MRobi1** [reply]
For me, using default compatibility, power works when on and off.
Switching to strong compatibility I see the opposite of you.
Shield on, power does not work.
Shield off, power does work and turns it on.

**Comment by MRobi1** [comment]
OK general testing observations. Testing done on both Shield TV Pro (2019) and Shield TV 2017. Results are identical with both.

Using "Default for SHIELD TV Pro (2019)" compatibility mode
D-Pad does not work, Play/Pause does not work, skip fwd and skip back does not work.
Power button DOES work.

Switching compatibility mode to "Strong"
All control buttons do work.
Power button does NOT work when device is on, but does work when device is off.

As mentioned above, YouTube button does not launch YouTube app.
I'd also like to see the ability to have less than 4 app launch buttons. Button 5 and 6 have a "- - - -". Buttons 1-4 do not have that option so an app needs to be selected. I figure many would like to keep it exactly like the original remote.

Also, I notice sometimes the home button is red, other times the play/pause button is red. I'm not really clear on why they're turning red?

**Reply by PRProd** [reply]
Buttons light up (turn red) when the device reports that certain states are active.
 - Power button will light up when the power is on, and go into a "dim" state when power is in "standby" mode
 - Home button (the circle on the shield remote - the home icon on a Fire remote) will light up when the device is on the home menu
 - Play/Pause button will light up when the device reports that the media player is in "playing" state.

I've noticed that the play/pause feature can be misleading in many cases.  Home assistant will show the device is "playing" not only when it's playing media, but also when other things are happening that I wouldn't personally consider "playing".  I've been toying with the idea of removing the lit play button, or at least allowing for it to be disabled in the YAML config.  I haven't decided yet.

**Reply by MRobi1** [reply]
You are correct, the play/pause feature does occasionally light up when nothing is playing. I like it with the power button.

**Comment by PRProd** [comment]
@jrspowers & @MRobi1, Thanks for your reports so far!  I'll be looking into everything you've mentioned.  If you find anything else, please let me know.  For now, I'm going to start a little checklist to make sure I don't forget anything that you've already mentioned.

- [x] Default compatibility mode does not seem to be defaulting properly, causing most buttons to fail
- [x] Power button is unreliable in "Strong" compatibility mode for Shield devices
- [x] Shield buttons do disappear when removed from YAML, but do not yet have a removal option through the GUI
- [x] YouTube button does not launch the YouTube app on Shield devices
- [x] Visible device name feature is not yet complete as it breaks the layout of the Shield remote
- [x] Original Shield remote style is not yet available, only the latest version
- [ ] Double press on home button should bring up the app switcher

**Reply by jrspowers** [reply]
that's very much done to test this & hopefully you can solve everything & that it will work nicely?  everything on the list is great!

**Reply by PRProd** [reply]
No need to worry @jrspowers, some of these are a lot more simple than they might look. :smile: 

I'll be checking items off the list as I work today.  I will wait until I have a bunch of things done before I publish the next beta for testing.

**Reply by jrspowers** [reply]
that's good and look forward to the next beta!

**Comment by PRProd** [comment]
[Beta 2](https://github.com/PRProd/HA-Firemote/releases/tag/v2.0.0-beta2) is available to download and test.

### Known issues still open/unresolved:
 - Visible device name feature is not yet complete as it breaks the layout of the Shield remote
 - Original Shield remote style is not yet available, only the latest version
 - Updated documentation is not yet written

**Comment by PRProd** [comment]
If anybody in this conversation also owns one or more Amazon Fire devices, I'd also appreciate testing with this new version on them as well.  I've modified a lot of the existing working Fire code to make it work in tandem with the new Shield code, so bugs might be lurking.

**Comment by MRobi1** [comment]
Just took Beta 2 for a spin. All buttons seem to be working both in default mode as well as strong mode. That includes powering on/off reliably. YouTube also launches just fine.

One thing I notice, and I don't know if possible... typically a double press of the home button will bring up the app switch screen. I'm not sure if that's possible since it uses adb commands and I'm sure the home and app switch are 2 separate commands. Registering a double press may not be possible unless it were a separate button.

**Reply by PRProd** [reply]
@MRobi1 - Thanks for testing!  Glad to hear that we are making positive progress!

Thanks for letting me know about the double press!  I didn't know that was a function that existed!  I'm not sure if I can capture that, but I'll sure give it a try!

**Reply by MRobi1** [reply]
If it can't be captured (I assume it can't), possibly a toggle in the settings to add an extra button on the remote for it. It would deviate slightly from the look of the original remote which is why I think it should be optional, but would at least allow for the functionality of it.

**Comment by PRProd** [comment]
[Beta 3](https://github.com/PRProd/HA-Firemote/releases/tag/v2.0.0-beta3) is available to download and test

### Known issues still open/unresolved:
 - Original Shield remote style is not yet available, only the newest version
 - Updated documentation is not yet written
 - Waiting to hear back from @80081 for more info on Now TV app https://github.com/PRProd/HA-Firemote/issues/49
 - Double press on home button should bring up the app switcher on the Shield

**Reply by MRobi1** [reply]
Double press on home button does not appear to be working for me through default or strong.

**Reply by PRProd** [reply]
It's not ready yet.

**Reply by MRobi1** [reply]
Oh woops, I read that backwards as if it was working LOL

**Reply by PRProd** [reply]
@MRobi1  - So... I'm not having the best luck on detecting double clicks and/or long press events through the language framework that I'm restricted to.  I'm stubborn, so I'm going to keep trying, but I don't want it to prevent me from launching the new version officially.  I just wanted to give you a heads up that I haven't forgotten about it.

**Comment by PRProd** [comment]
Has anyone found any new issues?  Can I assume that no news is good news?

**Reply by jrspowers** [reply]
I'm going to set test beta 3 later and then test I'm home around 9 pm Belgium time that's about another 2 hours!

**Reply by jrspowers** [reply]
@PRProd I just installed the new beta 3 & have rebooted but don't see any new app you added so don't know what went wrong?

**Reply by jrspowers** [reply]
@PRProd after downloading again the beta 3 works now test !

**Reply by PRProd** [reply]
Home Assistant sometimes has some very sticky cache.  Glad you thought to re-download!

**Reply by PRProd** [reply]
Other than the new apps, if you could confirm that the device name feature is still working as expected, that would be great.  

I'm going offline now for a day or so to celebrate my birthday with my family.

**Reply by jrspowers** [reply]
@PRProd  I can say after downloading again that they work very well, all buttons work perfectly, the new app addition also works perfectly, I have 1 more question, could there be 6 apps, then place 4 & I know that the remote will become larger, but  I don't think that's bad?  thanks in advance!

**Reply by jrspowers** [reply]
> @PRProd  I can say after downloading again that they work very well, all buttons work perfectly, the new app addition also works perfectly, I have 1 more question, could there be 6 apps, then place 4 & I know that the remote will become larger, but  I don't think that's bad?  thanks in advance!

@PRProd is this possible what do you think wall this would make it very good for me then I can use all the apps that I often use thanks in advance!

**Reply by PRProd** [reply]
@jrspowers - Yes it is possible.  I'll need some additional time to work on that, but I will add it back in.

**Reply by jrspowers** [reply]
> Yes it is possible. I'll need some additional time to work on that, but I will add it back in.

take the time you need I can now continue the work perfectly now but with those 2 extra apps it would be even better!

**Reply by PRProd** [reply]
Done.  You'll see it in the official release!

**Comment by jrspowers** [comment]
@PRProd @MRobi1 I do notice that it often delays or does not respond to the remote?  I can't really explain where this delay or non-functioning comes from so hopefully you can solve this?

**Reply by PRProd** [reply]
I wondered if it was a unique issue with my Home Assistant instance or my specific Shield device, but I've also noticed that sometimes the polling between Home Assistant and my Shield are slow or delayed too.  Since the Firemote card depends on that connection, it's probably why you are seeing an occasional issue.  Unfortunately, since that's not the Firemote card's fault, I can't do anything to help.

**Reply by jrspowers** [reply]
> I wondered if it was a unique issue with my Home Assistant instance or my specific Shield device, but I've also noticed that sometimes the polling between Home Assistant and my Shield are slow or delayed too. Since the Firemote card depends on that connection, it's probably why you are seeing an occasional issue. Unfortunately, since that's not the Firemote card's fault, I can't do anything to help.

yes I understand that well & I also experience this?  what I also notice is that it takes at least 1 minute for all those who switch apps to indicate on the remote itself & if it is pressed you can't do anything else for the first minute & sometimes I can't press an app myself to use remote buttons?

**Reply by PRProd** [reply]
I haven't needed to try it yet, but I wonder if restarting your home assistant hardware and/or restarting your Shield could help ease the delay for a while.

**Reply by jrspowers** [reply]
> I haven't needed to try it yet, but I wonder if restarting your home assistant hardware and/or restarting your Shield could help ease the delay for a while.

I can try that but that is for tomorrow wall here is now 23:50 I am going to bed quietly let me know something tomorrow

**Comment by PRProd** [comment]
@MRobi1 & @jrspowers ~  THANK YOU for your help!  I sincerely appreciate it!  I'm going to close this discussion for now, and work on pushing out the official release.  :tada:
